node/consensus/src/passthrough.rs

```diff
@@ -15,11 +20,37 @@ pub fn spawn_consensus(
     _out_gossip_tx: Sender<Gossip<ChainPayload>>,
     control_rx: Receiver<Control<BeaconWitnessSelector>>,
     consensus_tx: Sender<ChainConsensusBlockBody>,
-) {
-    // Even though we are not using control_rx we have to keep it alive for the duration of the
-    // passthrough consensus.
-    let task = payload_rx
-        .fold((consensus_tx, control_rx), |(consensus_tx, control_rx), p| {
+    batch_transactions: bool,
+) { 
+    // whether produce blocks immediately when receives a transaction
+    if !batch_transactions {
+        let task = payload_rx
+            .fold((consensus_tx, control_rx), |(consensus_tx, control_rx), p| {
+                let message: SignedMessageData<ChainPayload> = SignedMessageData {
+                    owner_sig: DEFAULT_SIGNATURE, // TODO: Sign it.
+                    hash: 0,                      // Compute real hash
+                    body: MessageDataBody {
+                        owner_uid: 0,
+                        parents: HashSet::new(),
+                        epoch: 0,
+                        payload: p,
+                        endorsements: vec![],
+                    },
+                };
+                let c = ChainConsensusBlockBody { messages: vec![message] };
+                tokio::spawn(consensus_tx.clone().send(c).map(|_| ()).map_err(|e| {
+                    error!("Failure sending pass-through consensus {:?}", e);
+                }));
+                future::ok((consensus_tx, control_rx))
+            })
+            .map(|_| ())
+            .map_err(|_| ());
+        tokio::spawn(task);
+    } else {
+        let message_queue = Arc::new(Mutex::new(vec![]));
```

Comment:
We can avoid creating a separate queue with a mutex by merging two streams: Interval and payload_rx. Using `merge` or `select` method of streams, then applying the same fold to batch payload together.